core/store/src/trie/prefetching_trie_storage.rs

```diff
@@ -386,19 +396,29 @@ impl PrefetchStagingArea {
                 if full {
                     return PrefetcherResult::MemoryLimitReached;
                 }
-                guard.slots.insert(key, set_if_empty);
+                guard.insert(key, set_if_empty);
+                self.notify_slots_update(guard);
                 PrefetcherResult::SlotReserved
             }
         }
     }
 
     fn clear(&self) {
-        self.lock().slots.clear();
+        let mut guard = self.lock_slots();
+        guard.clear();
+        self.notify_slots_update(guard);
+    }
+
+    /// This consumes locked mutex guard to make sure to unlock it before
+    /// notifying the condition variable.
+    fn notify_slots_update(&self, guard: MutexGuard<SizeTrackedHashMap>) {
+        std::mem::drop(guard);
```

Comment:
Is this actually a right thing to do? My intuition is that we want to notify first and unlock after, so that the mutex is fairly distributed along all the waiters on the condvar. If it is unlocked before the notification, some other waiter that’s not in a condvar might grab the mutex, leading to all the condvar waiters block immediately after them being notified (a-la thundering herd?)

The standard library doc [examples](https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html#examples) seem to corroborate with my understanding…

Reply:
FWIW I raised a question about this [upstream](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/condvar.20examples/near/316067795). They have some interesting insights :)